server.go

```diff
@@ -320,15 +1053,69 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
 	// or we purposefully deleted it.
 }
 
+// handleRetryPeerMsg deals with peers that have to be retried.  It is
+// invoked from the peerHandler goroutine.
+func (s *server) handleRetryPeerMsg(state *peerState, p *peer.Peer) {
+	if p == nil {
+		return
+	}
+
+	// Ignore new peers if we're shutting down.
+	if atomic.LoadInt32(&s.shutdown) != 0 {
+		srvrLog.Infof("Retry peer %s ignored - server is shutting "+
+			"down", p)
+		p.Shutdown()
+		return
+	}
+
+	retryDuration, ok := state.retries[p]
+	if !ok {
+		retryDuration = connectionRetryInterval / 2
+	}
+
+	srvrLog.Debugf("Retrying connection to %s in %s", p.Addr(), retryDuration)
+
+	// Start a goroutine to avoid blocking peerHandler when retrying peers.
+	go func() {
+		// Sleep for retryDuration and retry peer connection.
+		// If unsuccessful, repeat by increasing retryDuration.
+		select {
+		case <-time.After(retryDuration):
+			conn, err := btcdDial("tcp", p.Addr())
+			if err != nil {
+				srvrLog.Debugf("Failed to connect to %s: %v", p.Addr(), err)
+				retryDuration += connectionRetryInterval / 2
+				if retryDuration > maxConnectionRetryInterval {
+					retryDuration = maxConnectionRetryInterval
+				}
+				state.retries[p] = retryDuration
```

Comment:
This is going to race.  It's accessing state.retries unprotected form a separate goroutine.